nmdc_runtime/site/changesheets/scripts/missing_neon_soils_ecosystem_data.py

```diff
@@ -0,0 +1,338 @@
+            f"Searching for NMDC biosamples with omics processing {omprc_id}..."
+        )
+        nmdcbs_ids = omprc_to_bs_map.get(omprc_id, [])
+        logging.info(f"Found {len(nmdcbs_ids)} NMDC biosamples with {omprc_id}...")
+        for nmdcbs_id in nmdcbs_ids:
+            nmdcbs_req = runtime_api_client.request("GET", f"/biosamples/{nmdcbs_id}")
+            if nmdcbs_req.status_code != 200:
+                logging.error(
+                    f"Failed to retrieve NMDC biosample {nmdcbs_id}: {nmdcbs_req.status_code}"
+                )
+                continue
+            nmdcbs = nmdcbs_req.json()
+            nmdc_biosamples.append(nmdcbs)
+
+    logging.info(f"Found {len(nmdc_biosamples)} NMDC biosamples for {goldbs_id}...")
+    return nmdc_biosamples, omprc_records
+
+
+def compare_biosamples(goldbs, nmdcbs) -> List[ChangesheetLineItem]:
+    changesheet_line_items = []
+
+    # Check for missing ecosystem metadata
+    changesheet_line_items.extend(_check_ecosystem_metadata(goldbs, nmdcbs))
+
+    # Check for missing gold biosample identifiers
+    changesheet_line_items.extend(_check_gold_biosample_identifiers(goldbs, nmdcbs))
+
+    return changesheet_line_items
+
+
+def _check_ecosystem_metadata(goldbs, nmdcbs) -> List[ChangesheetLineItem]:
+    # nmdc to gold ecosystem key map
+    ecosystem_key_map = {
+        "ecosystem": "ecosystem",
+        "ecosystem_category": "ecosystemCategory",
+        "ecosystem_type": "ecosystemType",
+        "ecosystem_subtype": "ecosystemSubtype",
+    }
+    changesheet_line_items = []
+    for nmdc_key, gold_key in ecosystem_key_map.items():
+        if not gold_key in goldbs:
+            logging.warning(f"no {gold_key} for {goldbs['biosampleGoldId']}...")
+            continue
+        if not nmdc_key in nmdcbs:
+            changesheet_line_items.append(
+                ChangesheetLineItem(
+                    id=nmdcbs["id"],
+                    action="update",
+                    attribute=nmdc_key,
+                    value=goldbs.get(gold_key),
+                )
+            )
+            continue
+        if nmdcbs[nmdc_key] != goldbs.get(gold_key):
+            changesheet_line_items.append(
+                ChangesheetLineItem(
+                    id=nmdcbs["id"],
+                    action="update",
+                    attribute=nmdc_key,
+                    value=goldbs.get(gold_key),
+                )
+            )
+            continue
+
+    return changesheet_line_items
+
+
+def _check_gold_biosample_identifiers(goldbs, nmdcbs) -> List[ChangesheetLineItem]:
+    changesheet_line_items = []
+    goldbs_id = normalize_gold_id(goldbs["biosampleGoldId"])
+    if not goldbs_id in nmdcbs["gold_biosample_identifiers"]:
+        changesheet_line_items.append(
+            ChangesheetLineItem(
+                id=nmdcbs["id"],
+                action="insert",
+                attribute="gold_biosample_identifiers",
+                value=goldbs_id + "|",
```

Comment:
This reminds me: an existing value could contain a pipe character (I mean a string, in general, could contain one—not to say our schema would permit it to for a given field). I don't know whether our changesheet functionality provides a way to escape a pipe character. 🤷

Reply:
In this case we are only doing this with gold_biosample_identifiers, which have a defined format e.g. bs1123456 so I think this can stay as-is